mono/metadata/boehm-gc.c

```diff
@@ -2061,6 +2066,44 @@ mono_push_ephemerons (void)
 			if (!GC_is_marked (current_ephemeron->key)) {
 				mono_gc_wbarrier_generic_store_internal (&current_ephemeron->key, tombstone);
 				current_ephemeron->value = NULL;
+			}
+		}
+	}
+}
+
+static void
+mono_push_ephemerons (void)
+{
+	ephemeron_node* prev_node = NULL;
+	ephemeron_node* current_node = NULL;
+
+	/* iterate all registered Ephemeron[] */
+	for (current_node = ephemeron_list; current_node; current_node = current_node->next)
+	{
+		Ephemeron* current_ephemeron, * array_end;
+		MonoObject* tombstone = NULL;
+		/* reveal weak link value*/
+		MonoArray* array = REVEAL_POINTER (current_node->ephemeron_array_weak_link);
+
+		/* remove unmarked (non-reachable) arrays from the list */
+		if (!GC_is_marked (array)) {
+			continue;
+		}
+
+		prev_node = current_node;
+
+		current_ephemeron = mono_array_addr_internal (array, Ephemeron, 0);
+		array_end = current_ephemeron + mono_array_length_internal (array);
+		tombstone = array->obj.vtable->domain->ephemeron_tombstone;
+
+		for (; current_ephemeron < array_end; ++current_ephemeron) {
+			/* skip a null or tombstone (empty) key */
+			if (!current_ephemeron->key || current_ephemeron->key == tombstone)
+				continue;
+
+			/* If the key is not marked, then set it to the tombstone and the value to NULL. */
+			if (!GC_is_marked (current_ephemeron->key)) {
+
```

Comment:
Is this empty if statement correct or is it missing something here (line 2106)?

Reply:
It is intended to be empty. It allows the marking and clearing phases to have mirrored code. I could invert the `if` check and perhaps it would be clearer.

Reply:
Thanks Jonathan - Or just add a comment that it is intended :-)

Reply:
I adjusted the code to make it clearer. Thanks for taking a look @roberto-renzetti-unity